File: cillers/app/datastores/couchbase/models/cluster.py
import urllib
import time
from pathlib import Path
from datetime import timedelta
from couchbase.cluster import Cluster, ClusterOptions, ClusterTimeoutOptions
from couchbase.auth import PasswordAuthenticator
from couchbase.exceptions import CouchbaseException
from couchbase.management.buckets import BucketManager
from couchbase.options import WaitUntilReadyOptions
from couchbase.service_types import ServiceType
from config import config
from . import bucket, metadata

def ensure_ready_for_change(conf: config.ClusterChangeConfig):
    ensure_initialized(conf, 5*60)
    cluster = get(conf, 5*60)
    wait_until_ready(cluster, 5*60)
    metadata.ensure_initialized(cluster, conf)

def ensure_initialized(conf: config.ClusterChangeConfig, timeout_seconds: int):
    url = f"http://{conf.connection['host']}:8091/clusterInit"
    data = {
        'username': conf.credentials['username'],
        'password': conf.credentials['password'],
        'services': ','.join(services),
        'hostname': '127.0.0.1',
        'memoryQuota': '256',
        'sendStats': 'false',
        'clusterName': 'cillers',
        'setDefaultMemQuotas': 'true',
        'indexerStorageMode': 'plasma',
        'port': 'SAME'
    }
    encoded_data = urllib.parse.urlencode(data).encode()
    request = urllib.request.Request(url, data=encoded_data, method='POST')
    max_retries = 30
    for attempt in range(max_retries):
        try:
            response = urllib.request.urlopen(request, timeout=timeout_seconds)
            response_body = response.read().decode()
            print("Cluster initialization successful.")
            return
        except Exception as e:
            if attempt == max_retries - 1:
                print('Max retries exceeded')
                raise e
            error_message = str(e)
            if 'already initialized' in error_message or 'Unauthorized' in error_message:
                print("Cluster was already initialized.")
                return
            else: 
                print(f"The cluster is not responding. Retrying SHOULDN'T HAVE TO DO THIS ... {e}")
                time.sleep(1)
    assert False

def wait_until_ready(cluster: Cluster, timeout_seconds: int):
    print("Connecting to Couchbase ...")
    service_types = [
            ServiceType.Management,
            ServiceType.KeyValue,
            ServiceType.Query
        ]
    options = WaitUntilReadyOptions(service_types=service_types)
    for attempt in range(max_retries):
        try:
            cluster = get(conf, timeout_seconds)
            cluster.wait_until_ready(timeout_option, options)
            print("Couchbase is ready.")
            return
        except Exception as e:
            print(f"Retrying SHOULDN'T HAVE TO TO THIS  ... {e}")
            time.sleep(retry_delay)
    raise Exception("Failed to connnect to Couchbase.")

def get(conf: config.ClusterChangeConfig, timeout_seconds: int):
    max_retries = 20
    connection_string = generate_connection_string(conf)
    options = generate_options(conf)
    for attempt in range(max_retries):
        try:
            cluster = Cluster(connection_string, options)
            return cluster
        except Exception as e:
            if attempt == max_retries - 1:
                raise e 
            print(f"Cluster connection failed. Retrying ... Error: {e}")
            time.sleep(1)
    assert False

def generate_connection_string(conf: config.ClusterChangeConfig) -> str:
    return f"{conf.connection['protocol']}://{conf.connection['host']}"

def generate_options(conf: config.ClusterChangeConfig, timeout_seconds) -> ClusterOptions:
    timeout_options = ClusterTimeoutOptions(connect=timeout_seconds)
    auth_options = PasswordAuthenticator(conf.credentials['username'], conf.credentials['password'])
    return ClusterOptions(auth_options, timeout_options=timeout_options)


#def is_initialized(client_conf):
#    max_retries = 30
#    for attempt in range(max_retries):
#        try:
#            # If we can connect with credentials, the cluster has been initialized
#            cluster = get_cluster(client_conf)
#            cluster.buckets().get_all_buckets() 
#            return True
#        except CouchbaseException as e:
#            message = str(e)
#            if 'message=request_canceled' in message:
#                print("Cluster connection request was cancelled, retrying ...")
#                time.sleep(1)
#            elif 'message=authentication_failure' in message:
#                print("Failed to connect with credentials. We assume the cluster has not been initialized")
#                return False
#    raise Exception("Max retries exceeded")

File: cillers/couchbase-init/lib/controllers/cluster.go
package controllers

import (
    "fmt"
    "io"
    "net/http"
    "net/url"
    "strings"
    "time"

    "github.com/couchbase/gocb/v2"
)

type ClusterController struct {
    host     string
    username string
    password string
    tls      bool
    cType    string
}

func NewClusterController(host, username, password string, tls bool, cType string) *ClusterController {
    return &ClusterController{
        host:     host,
        username: username,
        password: password,
        tls:      tls,
        cType:    cType,
    }
}

func (c *ClusterController) GetType() string {
    return c.cType
}

func (c *ClusterController) getConnectionString() string {
    protocol := "couchbases"
    if !c.tls {
        protocol = "couchbase"
    }
    return fmt.Sprintf("%s://%s", protocol, c.host)
}

func (c *ClusterController) getClusterInitParams() (string, url.Values) {
    protocol := "https"
    port := "18091"
    if !c.tls {
        protocol = "http"
        port = "8091"
    }

    initURL := fmt.Sprintf("%s://%s:%s/clusterInit", protocol, c.host, port)
    data := url.Values{
        "username":            {c.username},
        "password":            {c.password},
        "services":            {"kv,n1ql,index,fts,eventing"},
        "hostname":            {"127.0.0.1"},
        "memoryQuota":         {"256"},
        "sendStats":           {"false"},
        "clusterName":         {"cillers"},
        "setDefaultMemQuotas": {"true"},
        "indexerStorageMode":  {"plasma"},
        "port":               {"SAME"},
    }
    return initURL, data
}

func (c *ClusterController) EnsureInitialized() error {
    initURL, data := c.getClusterInitParams()
    
    maxRetries := 100
    for attempt := 0; attempt < maxRetries; attempt++ {
        resp, err := http.PostForm(initURL, data)
        if err != nil {
            if attempt == maxRetries-1 {
                return fmt.Errorf("timeout: waiting until cluster is started: %v", err)
            }
            fmt.Println("Waiting until cluster is started ...")
            time.Sleep(time.Second)
            continue
        }
        defer resp.Body.Close()

        body, err := io.ReadAll(resp.Body)
        if err != nil {
            return fmt.Errorf("failed to read response body: %v", err)
        }

        responseStr := string(body)
        if strings.Contains(responseStr, "already initialized") || strings.Contains(responseStr, "Unauthorized") {
            fmt.Println("Cluster already initialized.")
            return nil
        }

        fmt.Println(responseStr)
        fmt.Println("Cluster initialization successful.")
        return nil
    }
    return fmt.Errorf("failed to initialize cluster after maximum retries")
}

func (c *ClusterController) Connect() (*gocb.Cluster, error) {
    opts := gocb.ClusterOptions{
        Authenticator: gocb.PasswordAuthenticator{
            Username: c.username,
            Password: c.password,
        },
    }
    if c.tls {
        opts.SecurityConfig.TLSSkipVerify = false
    }

    cluster, err := gocb.Connect(c.getConnectionString(), opts)
    if err != nil {
        return nil, fmt.Errorf("failed to connect to cluster: %v", err)
    }

    err = cluster.WaitUntilReady(time.Second*300, nil)
    if err != nil {
        return nil, fmt.Errorf("cluster not ready: %v", err)
    }

    return cluster, nil
}

func (c *ClusterController) ConnectWithRetry() (*gocb.Cluster, error) {
    maxRetries := 30
    retryInterval := time.Second

    var lastErr error
    for attempt := 0; attempt < maxRetries; attempt++ {
        cluster, err := c.Connect()
        if err == nil {
            return cluster, nil
        }
        lastErr = err

        if attempt == maxRetries-1 {
            return nil, fmt.Errorf("failed to connect after %d attempts: %v", maxRetries, lastErr)
        }

        fmt.Println("Waiting for connection to cluster ...")
        time.Sleep(retryInterval)
    }
    return nil, lastErr
}
